prov/gni/src/gnix_nic.c

```diff
@@ -50,6 +51,166 @@ static pthread_mutex_t gnix_nic_list_lock = PTHREAD_MUTEX_INITIALIZER;
 
 uint32_t gnix_max_nics_per_ptag = GNIX_DEF_MAX_NICS_PER_PTAG;
 
+/*******************************************************************************
+ * Helper functions.
+ ******************************************************************************/
+
+static inline struct gnix_tx_descriptor *
+		__desc_lkup_by_id(struct gnix_nic *nic,
+				     int desc_id)
+{
+	struct gnix_tx_descriptor *tx_desc;
+
+	assert((desc_id >= 0) && (desc_id < nic->max_tx_desc_id));
+	tx_desc = &nic->tx_desc_base[desc_id];
+	return tx_desc;
+}
+
+/*
+ * function to process GNI CQ RX CQES to progress a gnix_nic
+ * -note that the callback functions for the vc's are suppose to
+ * include the following call sequence to release the SMSG message
+ * and unlock the nic lock:
+ *
+	status = GNI_SmsgRelease(vc->ep);
+	fastlock_release(&nic->lock);
+ *
+ * The idea here is that is possible that the callback function
+ * may need to have some kind of post-nic-lock work to do.
+ * Want to avoid using recursive locks.
+ */
+
+static int __nic_rx_progress(struct gnix_nic *nic)
+{
+	int i, ret = FI_SUCCESS;
+	int vc_id = 0, have_lock = 1;
+	int max_id;
+	gni_return_t  status = GNI_RC_NOT_DONE;
+	gni_cq_entry_t cqe;
+	struct gnix_vc *vc;
+	uint8_t tag = GNI_SMSG_ANY_TAG;
+	void *msg_ptr;
+
+	/*
+	 * first see if there are any CQE's on the
+	 * tx cq. Calling GNI_CqTestEvent does not require
+	 * getting the nic lock
+	 */
+
+	status = GNI_CqTestEvent(nic->rx_cq);
+	if (status == GNI_RC_NOT_DONE)
+		return FI_SUCCESS;
+
+try_again:
+	fastlock_acquire(&nic->lock);
+	status = GNI_CqGetEvent(nic->rx_cq, &cqe);
+	if (status  == GNI_RC_NOT_DONE) {
+		fastlock_release(&nic->lock);
+		return FI_SUCCESS;
+	}
+
+	/*
+	 * no CQ overrun
+	 */
+	if (status == GNI_RC_SUCCESS) {
+		vc_id =  GNI_CQ_GET_INST_ID(cqe);
+		vc = __gnix_nic_elem_by_rem_id(nic, vc_id);
+		if (vc->conn_state != GNIX_VC_CONNECTED) {
+			vc->modes |= GNIX_VC_MODE_PENDING_MSGS;
+			ret = _gnix_vc_add_to_wq(vc);
+			goto out;
+		}
+
+		/*
+		 * what's going on here? The rule
+		 * is that there is not a one-to-one
+		 * relationship between CQEs and messages
+		 * in the mailbox, so need to poll on
+		 * the mailbox till there are no
+		 * more messages.
+		 */
+try_again2:
```

Comment:
Does there need to be so many goto statements? I understand using them for common error paths, but I think we should consider using static-inline functions for simple iteration constructs if the purpose of the goto is to remove an indentation level.